tests/integration-tests/tests/cloudwatch_logging/test_cloudwatch_logging.py

```diff
@@ -0,0 +1,461 @@
+            event_generator = (event for event in events if event["message"] == expected_tail)
+            assert_that(next(event_generator, None)).is_not_none()
+
+    def verify_log_group_created(self, log_groups):
+        """Verify whether or not the cluster's log group was created depending on whether it was enabled."""
+        assert_that_log_group_names = assert_that(log_groups).extracting("logGroupName")
+        if self.enabled:
+            assert_that_log_group_names.contains(self.log_group_name)
+        else:
+            assert_that_log_group_names.does_not_contain(self.log_group_name)
+
+    def verify_log_group_retention_days(self, log_groups):
+        """Verify whether or not the cluster's log group was created depending on whether it was enabled."""
+        if not self.enabled:
+            return  # Log group should not be created if not enabled.
+        log_group = next(
+            (group for group in log_groups if group["logGroupName"] == self.log_group_name),
+            None
+        )
+        assert_that(log_group).is_not_none().is_equal_to(
+            {"retentionInDays": self.retention_days}, include="retentionInDays")
+
+    def verify_agent_status(self, logs_state):
+        """Verify CloudWatch agent is running on the MasterServer (or not if not enabled)."""
+        expected_status = "running" if self.enabled else "stopped"
+        assert_that(logs_state).extracting("agent_status").contains_only(expected_status)
+
+    @staticmethod
+    def verify_logs_exist(logs_state):
+        """Verify that the log files expected to exist on the nodes of this cluster do."""
+        for host_dict in logs_state:
+            for log_path, log_dict in host_dict['logs'].items():
+                assert_that(log_dict).is_equal_to({'exists': True}, include='exists')
+
+    def run_tests(self, logs_state):
+        """Run all CloudWatch logging integration tests."""
+        log_groups = self.data_retriever.get_log_groups()
+        self.verify_log_group_created(log_groups)
+        self.verify_log_group_retention_days(log_groups)
+
+        LOGGER.info("state of logs for cluster:\n{}".format(json.dumps(logs_state, indent=4)))
+        self.verify_agent_status(logs_state)
+        self.verify_logs_exist(logs_state)
+
+        if self.enabled:  # Log streams are only relevant when the feature is enabled
+            observed_streams = self.data_retriever.get_log_streams(self.log_group_name)
+            expected_stream_index = self._get_expected_log_stream_index(logs_state)
+            self.verify_log_streams_exist(logs_state, expected_stream_index, observed_streams)
+            self.verify_log_streams_data(logs_state, expected_stream_index, observed_streams)
+
+
+@pytest.mark.parametrize("cw_logging_enabled", [True, False])
+@pytest.mark.regions(["us-west-1", "us-west-2", "us-east-1", "us-east-2"])
+@pytest.mark.instances(["t2.micro", "c4.xlarge"])
```

Comment:
let's just go with `t2.micro` for now

Reply:
The reason I included c4.xlarge is because in the Jenkins build_and_test pipeline, `c4.xlarge` is the smallest instance included in the the commercial and china instances lists. In looking at the pipeline again, my inclination is to enable it for c5.xlarge as well, given that's the only one in the default govcloud instances list.

I want these tests to be enabled in all three partitions. Do the build_and_test pipelines get launched with non-default instances lists on a regular basis?

Reply:
Not very often.

It's run on every instance type passed in, hence I would limit it to one instance type that runs across partitions. It looks like `c5.xlarge` is the best one to do that with.

Reply:
Okay. I'll include that one, and I'll keep `t2.micro` for the sake of manual testing without having to whitelist it first.